common/src/collision.rs

```diff
@@ -109,21 +178,140 @@ impl std::ops::Mul<&Ray> for na::Matrix4<f32> {
+        }
+
+        Some(VoxelAABB { bounds })
+    }
+
+    /// Iterator over grid points contained in the region, represented as ordered triples
+    pub fn grid_points(
+        &self,
+        axis0: usize,
+        axis1: usize,
+        axis2: usize,
+    ) -> impl Iterator<Item = (usize, usize, usize)> {
+        let bounds = self.bounds;
+        (bounds[axis0][0]..bounds[axis0][1]).flat_map(move |i| {
+            (bounds[axis1][0]..bounds[axis1][1])
+                .flat_map(move |j| (bounds[axis2][0]..bounds[axis2][1]).map(move |k| (i, j, k)))
+        })
+    }
+
+    /// Iterator over grid lines intersecting the region, represented as ordered pairs determining the line's two fixed coordinates
+    pub fn grid_lines(&self, axis0: usize, axis1: usize) -> impl Iterator<Item = (usize, usize)> {
+        let bounds = self.bounds;
+        (bounds[axis0][0]..bounds[axis0][1])
+            .flat_map(move |i| (bounds[axis1][0]..bounds[axis1][1]).map(move |j| (i, j)))
+    }
+
+    /// Iterator over grid planes intersecting the region, represented as integers determining the plane's fixed coordinate
+    pub fn grid_planes(&self, axis: usize) -> impl Iterator<Item = usize> {
+        self.bounds[axis][0]..self.bounds[axis][1]
+    }
+}
+
 #[cfg(test)]
-mod test {
-    use approx::assert_abs_diff_eq;
+mod tests {
+    use std::collections::HashSet;
+
+    use crate::dodeca::Vertex;
 
     use super::*;
 
+    /// Any voxel AABB should at least cover a capsule-shaped region consisting of all points
+    /// `radius` units away from the ray's line segment. This region consists of two spheres
+    /// and a cylinder. We only test planes because covered lines and points are a strict subset.
     #[test]
-    fn temp_shape_cast_example() {
-        // Directly hit the z=0 plane with a ray 0.5 units away.
-        let ray = math::translate_along(&na::Vector3::new(0.0, 0.0, -0.5))
-            * &Ray::new(math::origin(), na::Vector4::z());
-        assert_abs_diff_eq!(
-            shape_cast(na::Matrix4::identity(), &ray, 0.9).tanh_distance,
-            0.5_f32.tanh(),
-            epsilon = 1e-4
-        );
+    fn voxel_aabb_coverage() {
+        let dimension = 12;
+        let dual_to_grid_factor = Vertex::dual_to_chunk_factor() as f32 * dimension as f32;
+
+        // Pick an arbitrary ray by transforming the positive-x-axis ray.
+        let ray = na::Rotation3::from_euler_angles(0.1, 0.2, 0.3).to_homogeneous()
+            * math::translate_along(&na::Vector3::new(0.2, 0.3, 0.1))
+            * &Ray::new(na::Vector4::w(), na::Vector4::x());
+
+        let tanh_distance = 0.2;
+        let radius = 0.1;
+
+        let aabb = VoxelAABB::from_ray_segment_and_radius(
+            dimension,
+            dual_to_grid_factor,
+            &ray,
+            tanh_distance,
+            radius,
+        )
+        .unwrap();
+
+        let covered_planes: HashSet<_> = aabb.grid_planes(0).collect();
```

Comment:
`collect`ing this into a `HashSet` feels a bit silly when it starts out as a `Range` which is both smaller and faster to query against. Perhaps returning `impl Iterator` is excessively opaque in this case?

Reply:
I'm undecided on this. The advantage of `impl Iterator` is that it's consistent with the other `VoxelAABB` functions.

The speed disadvantage is minor, since it's still very fast, and tests need not be optimized unless they're slowing down CI.

If I test with lines or points, I would need a hashset.

Reply:
No sense overthinking support code in a test I suppose.